[C10D][BE]  Add exception handlers to c10d collectives function

wz337 opened the pull request:
Summary:
1. Add a decorator function @exception_handlers to  c10d collectives.
2. Update test(torch/distributed/distributed_c10d.py) to include mp tests for exception_handler.

```
python3 test/distributed/test_c10d_error_logger.py
```

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1307,6 +1309,27 @@ def batch_isend_irecv(p2p_op_list):
+            error_msg_dict = {
+                "func_name": f"{func.__name__}",
+                "args": f"{args}, {kwargs}",
+                "world_size": f"{get_world_size()}",
```
It looks like some tests are using the direct initialization python initialization of ProcessGroupGloo / ProcessGroupNCCL (https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_gloo.py#L208). This is technically unsupported as it skips the state dictionaries that we keep to track `default group`, `backend`, etc. which is set during `init_process_group`. However, since we have these tests in CI, we see failures like "Default process group has not been initialized, please make sure to call init_process_group." (e.g. https://github.com/pytorch/pytorch/actions/runs/3318590784/jobs/5482967896)

Rather than updating all the tests to use `init_process_group` in this PR. We should guard `get_world_size()`, `get_rank()`, and `get_backend()` with `is_initialized()`.

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -34,6 +35,7 @@
+from .c10d_error_logger import _get_or_create_logger
 from .rendezvous import register_rendezvous_handler, rendezvous  # noqa: F401
 
 __all__ = [
```
add "exception_handler" here to make public.

H-Huang reviewed (commented):
![image](https://user-images.githubusercontent.com/14858254/198102815-1478d579-ed80-438a-9e92-cc97952fd214.png)

To resolve these, update `__all__` in distributed_c10d.py with `exception_handler` and add `__all__ = []` to https://github.com/pytorch/pytorch/blob/master/torch/distributed/c10d_error_logger.py

Testing plan:
`python test/test_public_bindings.py -k test_correct_module_names`

H-Huang commented on test/distributed/test_c10d_error_logger.py:
```diff
@@ -1,17 +1,143 @@
+            self.assertEqual("broadcast", error_msg_dict["func_name"])
+
+            self.assertIn("args", error_msg_dict.keys())
+            self.assertEqual("(tensor([0, 1]), 5), {}", error_msg_dict["args"])
```
Seeing a failure for this test here: https://github.com/pytorch/pytorch/actions/runs/3325425880/jobs/5498283134. The issue is some CI machines use GPU and have only 2 devices, so the WORLD_SIZE becomes 3 due to your logic on line 31. You can either change this assert to do 
```python
self.assertEqual(f"(tensor([0, 1]), {WORLD_SIZE+1}), {{}}", error_msg_dict["args"])
```
or update WORLD_SIZE to always be 2 for all cases and hardcode the assert.

H-Huang commented on torch/distributed/c10d_error_logger.py:
```diff
@@ -11,23 +11,20 @@
 from torch.distributed.logging_handlers import _log_handlers
 
-_c10d_error_logger = None
+__all__ = ["_get_or_create_logger", "_get_logging_handler"]
```
I don't think we want to make these functions public, if you set __all__ to empty array then I think the test would still pass.

```suggestion
__all__ = []
```

wz337 replied:
Ye. It's in V14, forgot to amend and push earlier.

H-Huang commented on torch/distributed/logging_handlers.py:
```diff
@@ -9,6 +9,8 @@
 import logging
 from typing import Dict
 
+__all__ = ["_log_handlers"]
```
same comment as c10d_error_logger.py, `__all__ = []`

H-Huang reviewed (approved):
LGTM, feel free to merge when green

wz337 committed:
[C10D][BE] Add exception handlers to c10d collectives function (#87643)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/87643

1. Add a decorator function exception_handlers to  c10d collectives.
2. Update test(torch/distributed/distributed_c10d.py) to include mp tests for exception_handler.

```
python3 test/distributed/test_c10d_error_logger.py
```

Test Plan: Test in OSS.

Reviewed By: H-Huang

Differential Revision: D40281632

fbshipit-source-id: fbc8c9af6f6fd3b1141a47d0eab15bbafa9ecb39
